src/ck-core/cksyncbarrier.C

```diff
@@ -131,39 +164,33 @@ void CkSyncBarrier::propagate_atsync()
 
 void CkSyncBarrier::recvLbStart(int lb_step, int sourcenode, int pe)
 {
-  if (lb_step != cur_epoch || startedAtSync) return;
+  if (lb_step > currentKick)
+    currentKick = lb_step;
+  if (lb_step < cur_epoch || startedAtSync)
+    return;
+
   const int mype = CkMyPe();
   const int mynode = CkNodeOf(mype);
   if (sourcenode < mynode)
     received_from_left = true;
   else if (sourcenode > mynode)
     received_from_right = true;
   else if (rank0pe) // convert incoming pe number to local rank
-    rank_needs_flood[pe - mype] = false;
+    rank_needs_kick[pe - mype] = false;
   else
     received_from_rank0 = true;
-  if (clients.size() == 1 &&
-      clients.front()->chare->isLocMgr())  // CkLocMgr is usually a client on each PE
-    CheckBarrier(true);  // Empty PE invokes barrier on self on receiving a flood msg
+
+  if (clients.empty())
+    CheckBarrier();  // Empty PE invokes barrier on self on receiving a kick
 }
 
-void CkSyncBarrier::CheckBarrier(bool flood_atsync)
+void CkSyncBarrier::CheckBarrier()
 {
   if (!on) return;
 
   const auto client_count = clients.size();
 
-  if (client_count == 1 && !flood_atsync)
-  {
-    if (clients.front()->chare->isLocMgr()) return;
-  }
-
-  // If there are no clients, resume as soon as we're turned on
-  if (client_count == 0)
-  {
-    cur_epoch++;
-    CallReceivers();
-  }
+  if ((client_count == 0 && currentKick < cur_epoch) || startedAtSync) return;
```

Comment:
It would be good to retain the previous comment 'If there are no clients, resume as soon as we're turned on'. Also does currentKick < cur_epoch ensure that code does not return for a future epoch in case client count !=0 in the future? It would be to add a comment on that as well.

Reply:
That comment isn't strictly true anymore, now there's the additional stipulation that we must have also received a kick corresponding to the current epoch or beyond. In any case, I'll add more explanation.

`currentKick < cur_epoch` is used here to ensure that the barrier has received a kick corresponding to at least the current epoch. When there are no clients, this is the criterion that needs to be met in order to trigger the barrier. It has nothing to do with the client count in the future.

Reply:
More explanation added in updated commits.